typing/typecore.ml

```diff
@@ -2596,6 +2596,238 @@ let rec final_subexpression exp =
     -> final_subexpression e
   | _ -> exp
 
+let is_prim ~name funct =
+  match funct.exp_desc with
+  | Texp_ident (_, _, {val_kind=Val_prim{Primitive.prim_name; _}}) ->
+      prim_name = name
+  | _ -> false
+
+(* List labels in a function type, and whether return type is a variable *)
+let rec list_labels_aux env visited ls ty_fun =
+  let ty = expand_head env ty_fun in
+  if TypeSet.mem ty visited then
+    List.rev ls, false
+  else match get_desc ty with
+    | Tarrow (l, _, ty_res, _) ->
+        list_labels_aux env (TypeSet.add ty visited) (l::ls) ty_res
+    | _ ->
+        List.rev ls, is_Tvar ty
+
+let list_labels env ty =
+  let snap = Btype.snapshot () in
+  let result =
+    wrap_trace_gadt_instances env (list_labels_aux env TypeSet.empty []) ty
+  in
+  Btype.backtrack snap;
+  result
+
+
+(* Collecting arguments for function applications *)
+
+type untyped_apply_arg =
+  | Known_arg of
+      {
+        sarg : Parsetree.expression;
+        ty_arg : type_expr;
+        ty_arg0 : type_expr;
+        wrapped_in_some : bool;
+      }
+  | Unknown_arg of
+      {
+        sarg : Parsetree.expression;
+        ty_arg : type_expr;
+      }
+  | Eliminated_optional_arg of
+      {
+        ty_arg : type_expr;
+        level: int;
+      }
```

Comment:
As someone who is unfamiliar with the previous state of the codebase, it is not obvious what these things mean. Could you consider writing a few documentation comments that give concrete examples of `Known_arg`, `Unknown_arg`, and `Eliminated_optional_arg` arguments? Other questions I would naively ask myself are: what is the difference between `ty_arg` and `ty_arg0`, what `wrapped_in_some` means (maybe examples with and without), and what the `level` is.

Reply:
Wrapped in some corresponds to an optional argument application :
```ocaml
let f : ?x:_ -> _
let _ = f ~x
```
When typing the application, `x` will be stored in a `Known_arg` with `wrapped_in_some = true` if I understand correctly (I definitely agree that comments should be added at least for future contributors)